Send notification emails when contributors are inactive

iandunn opened the pull request:
Fixes most of #27, but doesn't send an email to the sponsoring company. That can be a future iteration while @angelasjin is working on the details of that.

This sends the email to folks that haven't logged in in 3 months. Once #119 and #189 are done, we can update it to look at 5ftF contributions specifically.

It needs polishing, but I think it's close to a state that I'm happy with, and close enough to get some review while I finish the polishing.

I initially tried using more complex SQL queries to reduce the number of rows that would be fetched, and the amount of PHP code needed, but that ended up being more trouble than it was worth.

See #205 for a discussion about the email copy.

iandunn committed:
Add stub for `natural_language_join()`.

iandunn committed:
Add cron job to send emails to inactive contributors.

iandunn committed:
Add email that is sent to inactive contributors.

tellyworth commented on plugins/wporg-5ftf/includes/email.php:
```diff
@@ -22,7 +23,7 @@
-function send_email( $to, $subject, $message, $pledge_id ) {
+function send_email( $to, $subject, $message, $pledge_id = false ) {
 	$headers = array(
 		'From: WordPress - Five for the Future <donotreply@wordpress.org>',
```
Some mixed messages here - we ask people to reply but then have `donotreply` as the From address. I realise the Reply-To will work, but some users might not.

iandunn replied:
That's a good point. I was thinking it was used in order to match the `Return-Path`, so that spam filters don't think we're forging `From`. I just checked though, and the `Return-Path` is `bounce@wordpress.org`.

We could change it to that. It technically has the same meaning, but most folks won't be familiar with the term, so it's not as bad.

I'm leery of changing the `From` to `getinvolved@` though, because of those spam filters.

iandunn replied:
I set it to `bounce@` in 7cad4bd, but I'm open to other ideas if you think there's a better way.

tellyworth commented on plugins/wporg-5ftf/includes/contributor.php:
```diff
@@ -566,3 +583,199 @@ function parse_contributors( $contributors, $pledge_id = null ) {
+	$inactivity_threshold = strtotime( INACTIVITY_THRESHOLD_MONTHS . ' months ago' );
+
+	foreach ( $contributors as $index => $contributor ) {
+		if ( $contributor['last_logged_in'] > $inactivity_threshold ) {
```
Does this need some logic to handle a missing/invalid `last_logged_in` value?

iandunn replied:
It should always exist at this point, it'll be `0` if nothing exists in the db.

https://github.com/WordPress/five-for-the-future/blob/8d6ea36d2fd998b4b18bd178198e83cd3f0141c5/plugins/wporg-5ftf/includes/contributor.php#L742

tellyworth commented on plugins/wporg-5ftf/includes/contributor.php:
```diff
@@ -566,3 +583,199 @@ function parse_contributors( $contributors, $pledge_id = null ) {
+ * Notify an inactive contributor.
+ */
+function notify_inactive_contributor( array $contributor ) : void {
+	if ( ! Email\send_contributor_inactive_email( $contributor ) ) {
```
Could there be a situation where sending fails repeatedly? If it's a permanent error like an invalid email address we'd effectively be silently retrying forever, if I'm reading correctly.

iandunn replied:
🤔 that's true. 

It looks like account registration and editing both validate the address with Core's `is_email()`. That's not RFC-compliant, but that might actually help us in this case, since IIRC there are lots of "valid" addresses that aren't routable. There's probably some edge cases though.

`wp_mail()` also uses `is_email()` to validate. So, I think an address will always either: 
1. be blocked by `wp_mail()` and therefore not trigger any spam filters that hurt our IP reputation; or
2. be sent to the address's mail server, and therefore get an updated `5ftf_last_inactivity_email`

In the case of `1`, though, that could maybe eventually cause some users to never get the email. If enough edge cases clumped together inside a batch group, then they could get stuck at the front of it, causing the others at the end of the group to always be ignored.

https://github.com/WordPress/five-for-the-future/blob/8d6ea36d2fd998b4b18bd178198e83cd3f0141c5/plugins/wporg-5ftf/includes/contributor.php#L596-L599

Adding a `shuffle()` before slicing would avoid that.

iandunn replied:
I updated 041abcf to include `shuffle()`

tellyworth commented on plugins/wporg-5ftf/includes/contributor.php:
```diff
@@ -566,3 +583,199 @@ function parse_contributors( $contributors, $pledge_id = null ) {
+	$field_names = array_flip( XProfile\FIELD_IDS );
+
+	foreach ( $user_xprofiles as $user ) {
+		$user->user_id = absint( $user->user_id );
```
`absint( $user->user_id )` makes me wonder what edge case this is intended for, and whether there's a better way to handle it.

iandunn replied:
IIRC it's just because the DB returns it as a string, but in other places we need to compare it to an int. We have to use strict comparison because the linter will complain if we don't. It also just makes me feel more comfortable, knowing that it's what I expect.

tellyworth commented on plugins/wporg-5ftf/includes/email.php:
```diff
@@ -205,3 +206,52 @@ function send_pledge_deactivation_email( $pledge ) {
+		return trim( $team );
+	}, $contributor['teams_names'] );
+
+	$message = sprintf( "
```
Should there be a last-chance check to make sure all the variables are set? Just to guard against silly errors like "you pledged 0 hours" and other things that are likely to wind up in a screenshot on social media :)

iandunn replied:
🤔 

* `$name` falls back to `user_nicename`, which should be guaranteed to exist
* `$contributor['hours_per_week']` and `team_names` have checks to make sure they're valid

https://github.com/WordPress/five-for-the-future/blob/8d6ea36d2fd998b4b18bd178198e83cd3f0141c5/plugins/wporg-5ftf/includes/contributor.php#L650-L651

https://github.com/WordPress/five-for-the-future/blob/8d6ea36d2fd998b4b18bd178198e83cd3f0141c5/plugins/wporg-5ftf/includes/contributor.php#L666-L669

It feels a bit [WET](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)/overkill to duplicate that in the email function, but I can see the concern.

iandunn committed:
Email: Set `From` header to `bounce@` to reduce confusion

Sometimes we do want folks to reply to messages, and the `donotreply` address contradicts that. `bounce@` has a similar meaning, but is less known. `bounce@` has the advantage of matching the `Return-Path`, which may reduce the chance of a message being flagged as spam for forging the `From` header.

See https://github.com/WordPress/five-for-the-future/pull/206#discussion_r938418815
